src/Core/Form/IdentifiableObject/DataProvider/ProductFormDataProvider.php

```diff
@@ -156,6 +175,45 @@ private function addShortcutData(array $productData): array
         return $productData;
     }
 
+    /**
+     * @param ProductForEditing $productForEditing
+     *
+     * @return array<string, mixed>
+     */
+    private function extractVirtualProductFileData(ProductForEditing $productForEditing): array
+    {
+        $data = [
+            'has_file' => false,
+        ];
+        $virtualProductFile = $productForEditing->getVirtualProductFile();
+
+        if (null !== $virtualProductFile) {
+            // @todo: some service to generate file path instead of doing it manually?
+            $filePath = $this->virtualProductFileDir . $virtualProductFile->getFileName();
+
+            $data = [
+                'has_file' => true,
+                'virtual_product_file_id' => $virtualProductFile->getId(),
+                'file' => [
+                    'file' => new File($filePath),
+                    'download_file_url' => $this->urlGenerator->generate(
+                        // @todo: implement new action instead of old one?
+                        'admin_product_virtual_download_file_action',
```

Comment:
I think it should be replaced at some point yes, since the purpose is to remove these old controllers later